### PR TITLE
Cursor & SelectMapper Refactor (WIP)

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request refactors the `SelectMapper` into separate raw and aggregate mappers and it refactors the `Cursor` implementation to return `int64` keys and `interface{}` values.

_This pull request is largely complete but still has test failures_

### CODE REVIEW

**Comment by otoolep** [comment]
Can you explain why this will help with the upcoming engine changes? We started months ago with a single Mapper type, the first round of DQ changes split it into Raw and Aggregate, and @dgnorton merged it back to a single type. Now this change splits it again.

I'd like to make sure I have the full context.

**Comment by dgnorton** [comment]
I didn't merge raw and agg mappers.  At least not that I remember.  :)

File: tsdb/engine/b1/b1.go [hunk 1]
```diff
@@ -549,32 +549,37 @@ type Tx struct {
 	engine *Engine
 }
 
-// Cursor returns an iterator for a key.
-func (tx *Tx) Cursor(key string, direction tsdb.Direction) tsdb.Cursor {
-	// Retrieve key bucket.
-	b := tx.Bucket([]byte(key))
+// Cursor returns an iterator for a key over a single field.
```
Comment (corylanou) [inline tsdb/engine/b1/b1.go]: I may be reading this wrong, but we say `single field` here, do we mean `single point`?  Later, in the `cursor.read` method we check to see if we have a single field or multiple fields, which implies we can cursor over multiple fields right?

**Comment by otoolep** [comment]
@dgnorton -- yeah, I stated that incorrectly. What you did do is embed the Raw Mapper inside a LocalMapper i.e. set a Local into Raw mode. In other words, we're churning the implementation of this, and I wanted to understand why we're reverting to an earlier form. Perhaps it's for the same reasons I first separated the two Mappers.

File: tsdb/engine/b1/b1.go (hunk 1 above)
Comment (benbjohnson) [inline tsdb/engine/b1/b1.go]: The `Cursor()` was to do a single field originally. It was too hard to split that out right now so we'll have to wrap the `pd` storage with a wrapping cursor to join individual field cursors.

**Comment by benbjohnson** [comment]
@otoolep I resolved some more test failures but it's really slow going. I'm getting a type assertion error now. It's difficult to keep track of which fields are getting processed where. I pushed my latest code (98328ab).

**Comment by otoolep** [comment]
OK @benbjohnson - I will take a look.

I was mistaken about work @dgnorton had done, I forgot that *I* merged the mappers.

File: .hooks/pre-commit [hunk 2]
```diff
@@ -7,7 +7,7 @@ if [ $fmtcount -gt 0 ]; then
 fi
 
 # Due to the way composites work, vet will fail for some of our tests so we ignore it
-vetcount=`go tool vet --composites=false ./ 2>&1  | wc -l`
+vetcount=`go tool vet --composites=false --methods=false ./ 2>&1  | wc -l`
```
Comment (corylanou) [inline .hooks/pre-commit]: :cry:

Comment (benbjohnson) [inline .hooks/pre-commit]: Let me know if you have a better method name than Seek().

It probably makes sense to move to iterators (with just a Next() method) instead of cursors but that's a refactor for another time.

File: cluster/shard_mapper.go [hunk 3]
```diff
@@ -146,8 +148,8 @@ func (r *RemoteMapper) Open() (err error) {
 	return nil
 }
 
-func (r *RemoteMapper) SetRemote(m tsdb.Mapper) error {
-	return fmt.Errorf("cannot set remote mapper on a remote mapper")
+func (r *RemoteMapper) SetRemote(m tsdb.Mapper) {
```
Comment (corylanou) [inline cluster/shard_mapper.go]: minor: I know it's outside the scope of this refactor but the fact that this func exists really bothers me.  I'm assuming it's only here to satisfy an interface, which again, doesn't feel right.

File: tsdb/cursor.go [hunk 4]
```diff
@@ -1,35 +1,22 @@
...
-}
+// EOF represents a "not found" key returned by a Cursor.
+const EOF = int64(-1)
 
-// Forward returns true if direction is reverse
-func (d Direction) Reverse() bool {
-	return d == Reverse
+// Cursor represents an iterator over a series.
+type Cursor interface {
+	Seek(seek int64) (key int64, value interface{})
```
Comment (corylanou) [inline tsdb/cursor.go]: minor: if we really don't want to collied with vet (and I'm not that torn about it) this could be `Find` maybe? or perhaps `SeekKey`.  The vet warning we had to disable isn't one that I really care a lot about as it doesn't prevent bugs, it's more of a style warning.

File: tsdb/cursor.go [hunk 5]
```diff
@@ -37,26 +24,27 @@ func (d Direction) Reverse() bool {
...
-	prev      []byte
-	direction Direction
+	cursors []Cursor
+	heap    cursorHeap
+	prev    int64 // previously read key
 }
 
 // Seek moves the cursor to a given key.
-func (mc *multiCursor) Seek(seek []byte) (key, value []byte) {
+func (mc *multiCursor) Seek(seek int64) (key int64, value interface{}) {
```
Comment (corylanou) [inline tsdb/cursor.go]: nit: we probably shouldn't have named returns here as we don't use them.

File: tsdb/cursor.go [hunk 6]
```diff
@@ -71,26 +59,32 @@ func (mc *multiCursor) Seek(seek []byte) (key, value []byte) {
...
+func (mc *multiCursor) Ascending() bool {
+	if len(mc.cursors) == 0 {
+		return true
+	}
+	return mc.cursors[0].Ascending()
+}
 
 // Next returns the next key/value from the cursor.
-func (mc *multiCursor) Next() (key, value []byte) { return mc.pop() }
+func (mc *multiCursor) Next() (key int64, value interface{}) { return mc.pop() }
```
Comment (corylanou) [inline tsdb/cursor.go]: nit: we probably shouldn't have named returns here as we don't use them.

File: tsdb/cursor.go [hunk 7]
```diff
@@ -71,26 +59,32 @@ func (mc *multiCursor) Seek(seek []byte) (key, value []byte) {
...
 
 // pop returns the next item from the heap.
 // Reads the next key/value from item's cursor and puts it back on the heap.
-func (mc *multiCursor) pop() (key, value []byte) {
+func (mc *multiCursor) pop() (key int64, value interface{}) {
```
Comment (corylanou) [inline tsdb/cursor.go]: nit: we probably shouldn't have named returns here as we don't use them.

File: tsdb/mapper.go [hunk 8]
```diff
@@ -360,671 +232,364 @@ func (lm *SelectMapper) nextChunkRaw() (interface{}, error) {
...
-func (pq pointHeap) Swap(i, j int) { pq[i], pq[j] = pq[j], pq[i] }
+	// Validate the fields and tags asked for exist and keep track of which are in the select vs the where
+	selectFields := mm.SelectFields(m.stmt)
+	selectTags := mm.SelectTags(m.stmt)
 
-func (pq *pointHeap) Push(x interface{}) {
-	item := x.(*pointHeapItem)
-	*pq = append(*pq, item)
-}
+	// If we only have tags in our select clause we just return
```
Comment (corylanou) [inline tsdb/mapper.go]: nit: this comment doesn't seem to line up with the current behavior.  We actually return an error.

File: cluster/shard_mapper.go (hunk 3 above)
Comment (benbjohnson) [inline cluster/shard_mapper.go]: The `RemoteMapper` is getting pulled out. I'll remove this.

File: tsdb/cursor.go (hunk 4 above)
Comment (benbjohnson) [inline tsdb/cursor.go]: I might change it to `SeekTo()` for now. It mostly bothers me that `io.Seeker` is in `go vet` when it's so rarely used.

File: tsdb/cursor.go (hunk 5 above)
Comment (benbjohnson) [inline tsdb/cursor.go]: Fixed.

File: tsdb/cursor.go (hunk 6 above)
Comment (benbjohnson) [inline tsdb/cursor.go]: Fixed.

File: tsdb/cursor.go (hunk 7 above)
Comment (benbjohnson) [inline tsdb/cursor.go]: Fixed.

File: .hooks/pre-commit (hunk 2 above)
Comment (otoolep) [inline .hooks/pre-commit]: I hit this before and just used the name `SeekTo`.

Comment (otoolep) [inline .hooks/pre-commit]: Ah, I see you also used `SeekTo`, according to a comment below. So can this `go vet` change be pulled?

File: tsdb/cursor.go (hunk 4 above)
Comment (otoolep) [inline tsdb/cursor.go]: Oh, you *might* change to `SeekTo`. :-)

File: tsdb/cursor.go [hunk 9]
```diff
@@ -71,26 +59,32 @@ func (mc *multiCursor) Seek(seek []byte) (key, value []byte) {
...
 	// Read items until we have a key that doesn't match the previously read one.
 	// This is to perform deduplication when there's multiple items with the same key.
 	// The highest priority cursor will be read first and then remaining keys will be dropped.
 	for {
-		// Return nil if there are no more items left.
+		// Return EOF marker if there are no more items left.
```
Comment (otoolep) [inline tsdb/cursor.go]: Why this change?

File: tsdb/engine/bz1/bz1_test.go [hunk 10]
```diff
@@ -1,38 +1,23 @@
...
+	"strings"
 	"testing"
-	"testing/quick"
 	"time"
 
 	"github.com/influxdb/influxdb/influxql"
 	"github.com/influxdb/influxdb/tsdb"
 	"github.com/influxdb/influxdb/tsdb/engine/bz1"
-	"github.com/influxdb/influxdb/tsdb/engine/wal"
+	//"github.com/influxdb/influxdb/tsdb/engine/wal"
```
Comment (otoolep) [inline tsdb/engine/bz1/bz1_test.go]: Remove?

File: tsdb/executor.go [hunk 11]
```diff
@@ -26,39 +26,6 @@ type Executor interface {
 	Execute() <-chan *influxql.Row
 }
 
-// Mapper is the interface all Mapper types must implement.
-type Mapper interface {
```
Comment (otoolep) [inline tsdb/executor.go]: Yeah, this should never really have been defined in this file, +1 on the move.

**Comment by otoolep** [comment]
I took a look at this, and it generally looks fine.

Can you give us a couple of lines on your thinking, about why you wanted this change?

File: tsdb/cursor.go (hunk 9 above)
Comment (benbjohnson) [inline tsdb/cursor.go]: `nil` isn't allowed now that we're using `int64` so I wanted to declare the marker more explicitly than simply doing `-1`.

File: tsdb/engine/bz1/bz1_test.go (hunk 10 above)
Comment (benbjohnson) [inline tsdb/engine/bz1/bz1_test.go]: Fixed.

**Comment by benbjohnson** [comment]
@otoolep Sure, the main goal for this change was to make it so that `Tx.Cursor()` could accept a field name instead of returning an encoded byte slice. That's because @pauldix's new engine is columnar-based and it handles encoding/decoding inside the engine. Unfortunately, breaking it into a single field was _really_ difficult so it accepts a list of fields (`[]string`) instead. We'll have to wrap Paul's cursors in a parent cursor to handle multiple fields being grouped for the short term.

There were a few issues with passing fields into the cursor:

1. Decoding had to move from `tsdb` down into the individual engines. This is how it should have been from the beginning so it's good that we made this change. It's ultimately the engine's responsibility to determine how to handle encoding.

2. Cursor and mapper code was very, very tightly coupled in many spots. This caused me to split out the aggregate, raw, and remote mappers out separately. These could still use some refactoring still but they're much cleaner than they were.

**Comment by otoolep** [comment]
Thanks @benbjohnson -- both of those points are good ones, with which I agree.
